### PR TITLE
Add support for client cert credential type

### PR DESCRIPTION
**Pull request by Zlaticanin**
This PR adds functionality to the `ClientCertificateGenerator` struct. Specifically, the `newClientCertificateGenerator` and `generate` methods have been implemented. These methods use a given configuration and an expiration date to create a client certificate. The `newClientCertificateGenerator` method sets default values if they are not provided in the configuration and validates the input parameters. The generate method generates the certificate using the configuration and the expiration date. 

More details in RFC [here](https://docs.google.com/document/d/1tayc7LxyDT8MOdgVFL9k7GGQeGCzFD5mnlLwHOyvObs/edit#)

**Dynamic Role Usage:**
```
vault write database-mongodbatlas/roles/my-dynamic-certificate-role \
    db_name=my-mongodbatlas-database \
    creation_statements='{"database_name": "$external", "x509Type":
	"CUSTOMER", "roles":
	[{"databaseName":"sample_training","roleName":"readWrite"}]}' \
    default_ttl="1h" \
    credential_type="client_certificate" \
    credential_config=ca_cert="$(cat path/to/ca_cert.pem)" \
    credential_config=ca_private_key="$(cat path/to/private_key.pem)" \
    credential_config=key_type="rsa" \
    credential_config=key_bits=2048 \
    credential_config=signature_bits=256 \
    credential_config=common_name_template="{{.DisplayName}}_{{.RoleName}}
```

PR on MongoDB Atlas Secrets Engine: https://github.com/hashicorp/vault-plugin-database-mongodbatlas/pull/52

### CODE REVIEW

File: builtin/logical/database/credentials.go
Comment (austingebauer) [inline builtin/logical/database/credentials.go:180]: I think we can remove this todo and leave it posed as a question on the PR. I considered templating of the entire subject but settled for common name since it was more simple.

File: builtin/logical/database/credentials.go
Comment (austingebauer) [inline builtin/logical/database/credentials.go:313]: We should find out if this is okay. Perhaps cryptosec can help here!

File: sdk/database/dbplugin/v5/grpc_client.go
Comment (austingebauer) [inline sdk/database/dbplugin/v5/grpc_client.go:124]: Wanted to note that we should manually test running the mongodb atlas plugin as an external plugin before merging this. I remember hitting some non-obvious errors when introducing new fields for the snowflake keypair credential type. Happy to help with this!

File: builtin/logical/database/credentials.go
Comment (cipherboy) [reply builtin/logical/database/credentials.go:313]: What database is this for? I think the short answer is it depends on the auth model of the database itself: does it limit permissions based on other subject attributes?

File: builtin/logical/database/credentials.go
Comment (fairclothjm) [inline builtin/logical/database/credentials.go:356]:
```suggestion
// configMap returns the configuration of the ClientCertificateGenerator
```

File: builtin/logical/database/credentials.go
Comment (Zlaticanin) [reply builtin/logical/database/credentials.go:311]: @cipherboy This is for MongoDB Atlas secrets engine. Here is the PR that's on the db side: https://github.com/hashicorp/vault-plugin-database-mongodbatlas/pull/52

Comment (cipherboy) [reply builtin/logical/database/credentials.go:311]: Right, but how is MongoDB Atlas _doing_ the authentication? What attributes are there that it is validating against? Are there custom extension that are important? Or is it just "any certificate signed by a given CA is accepted"? &c.

Happy to talk on a call if you'd like to discuss it more. :-)

Comment (austingebauer) [reply builtin/logical/database/credentials.go:311]: I can explain some of the details around how MongoDB Atlas is doing the authentication. Might be easier over a quick call if you don't mind. I'll set something up!

File: builtin/logical/database/credentials.go
Comment (cipherboy) [inline builtin/logical/database/credentials.go:323]: I'd leave this to `false`. There's  no real reason to set a BasicConstraint with `IsCA=false` here, as that's very clearly the default behavior per spec, unless something breaks and requires it (though, they are outside of IETF RFC 5280 and likely ITU X.509 if so).

File: builtin/logical/database/credentials.go
Comment (cipherboy) [inline builtin/logical/database/credentials.go:324]: Might be worth a non-zero value here to back date it just slightly (`30s` is the default in PKI) just in case times are out of sync between Vault, the client, and MongoDB.

File: builtin/logical/database/credentials.go
Comment (cipherboy) [inline builtin/logical/database/credentials.go:325]: As below, I think this accepts a `nil` value which means it could be elided altogether IIRC?

File: builtin/logical/database/credentials.go
Comment (cipherboy) [inline builtin/logical/database/credentials.go:220]: Just since I saw it explicitly listed in the example on the call, I'd default to `0` for auto-detection here in your API. Nobody should care about this parameter much, outside of very few users, and I'd even perhaps suggest eliding it altogether until someone explicitly requests it tbh.

File: builtin/logical/database/credentials.go
Comment (cipherboy) [inline builtin/logical/database/credentials.go:260]: Strictly I think you should also validate that `BasicConstraintsValid` is asserted, too: https://pkg.go.dev/crypto/x509#Certificate

File: builtin/logical/database/credentials.go
Comment (cipherboy) [reply builtin/logical/database/credentials.go:311]: This should definitely be ok from what I saw. Just note that `OU/O/DN` might need to be set, depending on what the `tlsListener` had set on the MongoDB side :-)

**Review by cipherboy — APPROVED** [review]
Looks good! Thank you both @austingebauer and @Zlaticanin for the walk through on the call :)

File: builtin/logical/database/credentials.go
Comment (Zlaticanin) [reply builtin/logical/database/credentials.go:179]: Nope, none of these actually need to be exported. They are just used within the database secrets engine itself. I will drop the rest to use the lowercase as well.

File: builtin/logical/database/credentials.go
Comment (austingebauer) [reply builtin/logical/database/credentials.go]: Ty! Set `URLs` to nil in https://github.com/hashicorp/vault/pull/20425/commits/1a6969596e55dd5e01353ebcae0e774295ebb81a.

File: builtin/logical/database/credentials.go
Comment (austingebauer) [reply builtin/logical/database/credentials.go]: Done in https://github.com/hashicorp/vault/pull/20425/commits/3e22a33ac4fe012634d2ba58003b1282b13baf48. Thanks!

File: builtin/logical/database/credentials.go
Comment (austingebauer) [reply builtin/logical/database/credentials.go]: Done in https://github.com/hashicorp/vault/pull/20425/commits/3e22a33ac4fe012634d2ba58003b1282b13baf48. Thanks!

File: builtin/logical/database/credentials.go
Comment (austingebauer) [reply builtin/logical/database/credentials.go]: Done in https://github.com/hashicorp/vault/pull/20425/commits/328929264668bacf82d6802ac5b456cc5f5c6672. Thanks!

File: builtin/logical/database/credentials.go
Comment (austingebauer) [reply builtin/logical/database/credentials.go:220]: This is what we're doing right now for setting the default signature bits when 0 is provided via the API: https://github.com/hashicorp/vault/blob/328929264668bacf82d6802ac5b456cc5f5c6672/builtin/logical/database/credentials.go#L284-L307. Could you advise on if this look correct or if there was some other way to auto-detect that you had in mind? Thanks!

File: builtin/logical/database/credentials.go
Comment (austingebauer) [reply builtin/logical/database/credentials.go:260]: Thanks! Done in https://github.com/hashicorp/vault/pull/20425/commits/02ef08ed3f8f74e54b60d0020caebfbb63281feb.

File: builtin/logical/database/credentials.go
Comment (austingebauer) [reply builtin/logical/database/credentials.go:293]: Good to know! We currently don't have a default for the key type and require it as input. If you were to recommend a default for this, which one would you suggest? 🤔

File: builtin/logical/database/credentials.go
Comment (cipherboy) [reply builtin/logical/database/credentials.go:224]: Aha, I was thinking this function: `ValidateDefaultOrValueKeyTypeSignatureLength`

File: builtin/logical/database/credentials.go
Comment (cipherboy) [reply builtin/logical/database/credentials.go:264]: Ah, I suppose this is equivalent. :laughing: I'm used to reading:

```
	if !parsedBundle.Certificate.BasicConstraintsValid || !parsedBundle.Certificate.IsCA {
	...
	}
```

but this works :-)

File: builtin/logical/database/credentials.go
Comment (cipherboy) [reply builtin/logical/database/credentials.go:297]: I'd perhaps suggest `ec/256`, its fast and should be reasonably well supported.

File: builtin/logical/database/credentials.go
Comment (cipherboy) [reply builtin/logical/database/credentials.go:224]: Something makes me think you could pass through `0` as signatureBits fwiw. Not tried it recently though :-D

File: builtin/logical/database/credentials.go
Comment (Zlaticanin) [reply builtin/logical/database/credentials.go:299]: done in [801daa8](https://github.com/hashicorp/vault/pull/20425/commits/801daa824e866de7461855a33074ba4ef769f108)

File: builtin/logical/database/credentials_test.go
Comment (raymonstah) [inline builtin/logical/database/credentials_test.go:268]: nit: It would be good to specific about the exact error instead of just the presence of an error.

**Review by raymonstah — APPROVED** [review]
Just a minor comment about testing for specific errors, but ✅

File: builtin/logical/database/credentials.go
Comment (austingebauer) [reply builtin/logical/database/credentials.go:226]: Thanks! It does look like you can. We'll take a look to see if we can use this.

File: changelog/20425.txt
Comment (austingebauer) [inline changelog/20425.txt:2]: There is a [special format](https://hashicorp.atlassian.net/wiki/spaces/VAULT/pages/1311244491/Changelog+Process#New-and-Major-Features) for features in the changelog. 

Suggest something like:
```suggestion
**MongoDB Atlas Database Secrets**: Adds support for client certificate credentials
```

File: builtin/logical/database/credentials.go
Comment (austingebauer) [inline builtin/logical/database/credentials.go:225]: We should set the default signature bits too when 0 is provided. Let's see if we can use the function recommended by Alex.

File: builtin/logical/database/credentials.go
Comment (vinay-gopalan) [inline builtin/logical/database/credentials.go:287]: nit: is it possible to write a test for this? Understand if it is hard to do with the SDK 🙏🏼

**Review by vinay-gopalan — APPROVED** [review]
Looks great! Had 1 nit about potentially adding a test, but LGTM otherwise. Great work! 🎉

**Review by austingebauer — APPROVED** [review]
LGTM. Great job, @MilenaHC 🚀

File: builtin/logical/database/credentials.go
Comment (austingebauer) [inline builtin/logical/database/credentials.go:179]: We can create a separate task for this, but it would be good to add comments to each of the exported fields here to indicate if they're required/optional and what values can be provided to them (similar to what's done in `rsaKeyGenerator`).

**Review by austingebauer — APPROVED** [review]

File: builtin/logical/database/credentials.go
Comment (Zlaticanin) [reply builtin/logical/database/credentials.go:179]: done in [0abd0a4](https://github.com/hashicorp/vault/pull/20425/commits/0abd0a450a0cd0e2e7723659150ab349a0061002)

File: builtin/logical/database/credentials.go
Comment (Zlaticanin) [reply builtin/logical/database/credentials.go:303]: I'll create a separate task for this, it will require some more time

**Pull request merged**
